### ISSUE TITLE
Test-Json doesn't recognize JSON arrays or primitives

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#11384]
# Steps to reproduce

```powershell
'42' | Test-Json | Should -BeTrue
'[ 1, 2 ]' | Test-Json | Should -BeTrue
```

# Expected behavior

The tests should succeed, given that these are valid [JSON strings](https://www.json.org/json-en.html) and that piping them to `ConvertFrom-Json` works just fine:

```powershell
# These tests pass.
'42' | ConvertFrom-Json | Should -Be 42
'[ 1, 2 ]' | ConvertFrom-Json | Should -Be 1, 2
```

# Actual behavior

Both tests fail.

# Environment data

```none
PowerShell Core 7.0.0-rc.1
```

**Comment by iSazonov** [comment]
@mklement0 Could you please add more info why it should work in the way? Maybe reference to a standard or a quote from the standard. It would help to avoid regressions in follow fix.

**Comment by mklement0** [comment]
@iSazonov, I've added a link to json.org to the OP, and also a demonstration that `ConvertFrom-Json` itself happily accepts such inputs.

**Comment by iSazonov** [comment]
We use NewtonSoft JObject.Parse() which throws. I tried System.Text.Json.JsonDocument.Parse() and … most of our tests failed because examples in the issue and out tests in Test-Json.Tests.ps1 use single quotes but Json standard allows only double quotes.
So we should decide what to do because switching to JsonDocument.Parse() looks like a breaking change. 
With double quotes the tests and the issue examples are passed if JsonDocument.Parse() is used.

**Comment by mklement0** [comment]
> use single quotes but Json standard allows only double quotes.

The single quotes aren't in the _data_, they are only there as _PowerShell's string delimiter_.

What `Test-Json` sees is just `42` and `[ 1, 2 ]` and both are valid JSON strings.

I haven't looked at the internals, but it's obvious that whatever `ConvertFrom-Json` manages to parse should be considered valid by `Test-Json`.

**Comment by jochenvanwylick** [comment]
+1 Encountered the same problem here - would love to see this resolved!

**Comment by iSazonov** [comment]
@jochenvanwylick You could download artifact from #11397, check your scenarios and report issues there.

**Comment by mklement0** [comment]
@iSazonov: In the current implementation,

https://github.com/PowerShell/PowerShell/blob/7b33cfe8470e100e4de3942eecbb57549e293747/src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs#L70

should be:

```c#
parsedJson = JToken.Parse(Json);
```

Only `JToken` correctly parses JSON documents such as `42`  and `[ 1, 2 ]`, whereas `JObject` accepts just a _single object_ (dictionary).

Or, better yet, use the same technique that `ConvertFrom-Json` uses:

https://github.com/PowerShell/PowerShell/blob/7b33cfe8470e100e4de3942eecbb57549e293747/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L179-L187

Note that the code there contains a workaround for a NewtonSoft bug that's no longer needed.

--- 

In a `System.Text.Json`-based implementation, `System.Text.Json.JsonDocument.Parse()` is indeed the right tool.

---

All these tests succeed:

```powershell
# Valid JSON
'42', '[ 1, 2 ]', '{ "foo": 1 }' | % {
    {  [System.Text.Json.JsonDocument]::Parse($_) } | Should -Not -Throw
    {  [Newtonsoft.Json.Linq.JToken]::Parse($_)  } | Should -Not -Throw
}

# Invalid JSON
'a', '[', '{ "foo": 1 }, { "bar": 2 }' | % {
    {  [System.Text.Json.JsonDocument]::Parse($_) } | Should -Throw
    {  [Newtonsoft.Json.Linq.JToken]::Parse($_)  } | Should -Throw
}

```

**Comment by iSazonov** [comment]
@mklement0 Thanks! I think we should move to Core API if possible. Sooner or later it will happen anyway. If there are no serious problems, it is better to do it "sooner".

**Comment by thedavecarroll** [comment]
Has there been any progress on this?

A Redditor recently made a [similar comment](https://www.reddit.com/r/PowerShell/comments/fgopt5/powershell_7_changes_to_json_cmdlets/fk6y8hh?utm_source=share&utm_medium=web2x). I responded with a link to this issue.

Basically, I said that they could use `ConvertFrom-Json | ConvertTo-Json | Test-Json` to perform tests. 

```powershell
'[ { name : "Joe" } ]' | ConvertFrom-Json | ConvertTo-Json | Test-Json
```

Is this the best advice for **PowerShell 7 GA**?

**Comment by mklement0** [comment]
@thedavecarroll, yes, it's unfortunate that this wasn't addressed in time for 7.0

PR #11397 will implicitly fix these issues, if I understand correctly, but due to its move to a different underlying API it will bring breaking changes, notably no longer  being able to use _single_-quoted JSON (right, @iSazonov?).

I wasn't even aware of the other issue that `Test-Json` doesn't support its input as an _array_ of lines, the way `ConvertFrom-Json` does  (e.g., `'{ "foo": ', '"bar" }' | Test-Json` fails, which succeeds with `ConvertFrom-Json`)

As for a workaround: It's sufficient to use only `ConvertFrom-Json`, and derive from its failure whether the input is valid JSON or not:

```powershell
$valid = try { '[ { name : "Joe" } ]' | ConvertFrom-Json; $true } catch { $false }
```

**Comment by iSazonov** [comment]
@mklement0 Yes. 
Currently we are waiting .Net 5.0 public preview to continue.

**Comment by dgard1981** [comment]
Might we get another update on this issue?

@iSazonov said -
> Currently we are waiting .Net 5.0 public preview to continue.

Looks like this has [now happened](https://devblogs.microsoft.com/dotnet/announcing-net-5-0-preview-5/).

With regards to suggested workarounds using `ConvertFrom-Json` - unfortunately that's no good if you are trying to test against a schema.

**Comment by iSazonov** [comment]
We have PRs for ConvertFrom-Json and Test-Json a long time without review. Unfortunately, we missed this time. I think it makes no sense to continue this now. .Net 5 will be frozen soon for release and we have no chance to fix something there if we find problems.

**Comment by heilkn** [comment]
Hello,

what does that mean now? The world will be stuck with a broken command? Could you please make a clear statement, whether this issue will be abandoned or resolved?

Kind regards

Konstantin

**Comment by gogbg** [comment]
I'm encountering the same issue.
Please fix it, before .net 5.0

**Comment by MalcolmTyrrell** [comment]
In case it's useful to anyone, I've had some success working around this issue with code like the following:
```
try
{
    # Work around a limitation of Test-Json, which can't parse top-level arrays:
    # Wrap the actual json as the value of a single property "_" in an object.
    $schemaPrefix = "{`"type`":`"object`",`"properties`":{`"_`":"
    $schemaBody = $schema -replace "`"\`$ref`"\s*:\s*`"#/", "`"`$ref`":`"#/properties/_/"
    $schemaSuffix = "}}"
    $wrappedSchema = $schemaPrefix + $schemaBody + $schemaSuffix
    $wrappedJson = "{`"_`":" + $json +"}"
    Test-Json -Json $wrappedJson -Schema $wrappedSchema -ErrorAction Stop
    return $True
}
catch
{
    # Have to unwrap the path in any schema error
    $schemaError = $_.ErrorDetails.Message.Replace("#/_.", "#/").Replace("#/_[", "#/[")
    Write-Error "$($_.Exception.Message) $($schemaError)"
    return $False
}
```

**Comment by iSazonov** [comment]
You could try artifact from #11397 (now it is on 5.0 RC2) - both tests from the OP are passed.

> single-quoted JSON

This was rejected by .Net team because of (1) low audience, (2) performance lack. https://github.com/dotnet/runtime/issues/31608 It is true for Test-Json too. I believe we have no need to think about this today.

**Comment by hacst** [comment]
Also stumbled over this. I reported a bug against the documentation of Test-Json so we can at least have the limitations documented. As it is now Test-Json does not really validate JSON. It accepts invalid JSON in the form of single quotes and comments while rejecting valid JSON like arrays.

While I would prefer it to actually validate JSON I'm fine with it working with a dialect and being limited to objects if that can no longer be changed. But at least it should be clearly documented.

**Comment by Michaelvsk** [comment]
Seems that it hasen't been fixed up until now? Just tried with `PowerShell 7.2.0-preview.8` and still same problem (array not considered a valid JSON). :(

**Comment by iSazonov** [comment]
No, PR is waiting 1.5 years.

**Comment by Szeraax** [comment]
After having read through all of these comments, I'm convinced that MS wants you to use try/catch for validating json. And now I think I have seen this before. So sad that we will have to keep waiting on this bug to get resolved. 

Imagine if Test-Path/Resolve-Path and Get-Item/gci worked differently for hidden items......... (sarcasm, *Item cmdlets don't find hidden paths, but *-Path cmdlets do).

**Comment by JordanTheITGuy** [comment]
This is pretty frustrating. 

This basically means that if you want to do anything more complicated than a single object you have to break the array into unique objects, which is terrible from a performance perspective, and otherwise you're stuck with try-catch stuff.

**Comment by traberc** [comment]
I just stumbled across this issue and am baffled that it hasn't been resolved in 2 years, even in the latest versions of PowerShell.

I am currently running PowerShell 7.2 in Visual Studio Code and `Test-Json` reports a JSON file is invalid with an array as the root object. Both Visual Studio Code and `ConvertFrom-Json` treat this file as valid JSON. 

As others have mentioned, regardless of what JSON standard PowerShell is conforming to, `Test-Json` and `ConvertFrom-Json` should treat the same input as either valid or invalid JSON. They should not give different results.

In my case, I can change the schema of my source config file, but that's a lame workaround for a 2-year-old bug.

**Comment by m-soltani** [comment]
Really frustrating that this issue still exists. I have exactly a case for which I can not develop a workaround and `Test-Json` cmdlet should really have this issue fixed.
Anybody have a working solution when you have to really test against a json schema ?

**Comment by MalcolmTyrrell** [comment]
If you're desperate, you could try the hacky code in my comment above. It seemed to work OK for a scenario I had.

**Comment by hilari0n** [comment]
Yes, this is unbelievable and frustrating, that such a simple and obvious error is still not fixed.
I ended up writing my own Test-Json script. Unfortunately I can't share it, as it's now the property of my employer.

Looking at the original code it's not that hard to write your own. You can safely assume, that Newtonsoft.Json and NJsonSchema libraries are present, as PowerShell uses them for this crippled implementation, so they are a mandatory dependency. So you can parse your Json using Newtonsoft.Json, parse your schema using NJsonSchema and do the validation of the first by the second, so using the same libraries the Test-Json command does and probably a similar approach.
You can add your own features too, e.g. accept Json file paths (not just Json content) and maybe with wildcards, load Json schema indicated in the Json document, provide a way to add PowerShell code validation (e.g. where Json schema can't cover some requirements), change the output format (so you can have results for multiple Json files and it's clear which error is for which document), allow some deserialization options to be set (like max depth, date or number processing), the parsed Json to be returned (so you do not have to parse it separately), etc.
Most of this can be done while maintaining compatibility with the Test-Json command (i.e. it failing for non-objects), by having the needed behavior work on switches and keeping the default call perform the same crap as Test-Json.

**Comment by eric-ash-1** [comment]
I would also love to see this bug fixed. It's still there in 7.2.3
